### DISCUSSION TITLE
Collection Literals - zero-copy construction for user-defined types

### DISCUSSION THREAD

**Discussion by neon-sunset** [dotnet/csharplang#7499] [Language Ideas]
## Context

Currently, it is not possible to construct a type in a way that would allow to skip copying contents to an intermediate buffer, which is then consumed by `.Create` collection builder method in the form of `ReadOnlySpan<byte>`.

This becomes especially problematic in performance-sensitive scenarios which may involve aribtrarily large collections or when implementing various `ImmutableArray`-like primitives.

For example:
```C#
U8String text = [(byte)'#', (byte)':', ..payload];
```
Here, the user-defined type is backed by `byte[]` and construction through allocating (or renting) an intermediate buffer, then consuming it via `.Create(ROS<byte>)` only to copy it to a newly allocated `byte[]` leads to an undesirable overhead.

There exist other scenarios, which would not be covered by explicit support for `Create(T[])` either:
```cs
// Allocates and then frees a NativeMemory-backed buffer
using NativeString text = [(byte)'{', ...payload, (byte)'}']; 
```

## Design
The following designs were mentioned in `csharp-lang-design` discord channel that could address zero-copy construction:

### Option 1
`.Create<T>(T[] value)`

Pros:
- Compatible with async in its current form
- Very simple design for constructing array-backed types

Cons:
- `T[] value` must be passed directly to `.Create` and otherwise not exposed to user code.
This might require either allowing an `internal` method to be a target of `CollectionBuilderAttribute` or allowing it to target a builder method declared on a different type (like `UserTypeMarshal`) visible in the current assembly (if not already)
- Offers limited benefits to types not backed by arrays

### Option 2
`Create<T>(int length, out Span<T> span)`

Pros:
- Only assumes that the constructed collection is backed by contiguous memory
- Not dissimilar to the currently existing patterns for (unsafely) constructing `List<T>` with `.SetCount` + `.AsSpan`

Cons:
- Not compatible with async (without relaxing the restriction to allow `ref struct`s to exist between await/yield points)
- Potentially more error-prone design?

As was noted in `csharp-lang-design`, both options could coexist to cover all scenarios.

_Submitting this discussion as was suggested by @jnm2_

**Comment by jnm2** [comment]
Thanks for these motivating scenarios! We want to start planning for these in C# 13 after dictionaries and natural types, but conversations related to this are already happening.

For instance, there's been discussion about automatically calling implicit conversion operators from array, so that any type that is implicitly convertible from T[] (or possibly span) is itself constructable via that implicit conversion. But U8String does not have that implicit conversion (as you confirmed) because converting from a byte array might throw for invalid UTF8 sequences, and throwing would be inappropriate during implicit conversions. So, U8String still motivates the construction patterns you outlined above.

Everyone, please chime in with examples of user-defined collection types like this, where neither the `Create<T>(ReadOnlySpan<T>)` route nor the collection initializer route for mutable user-defined collection types (repeated `Add(T)` calls) performs as well as you'd like.

**Comment by TahirAhmadov** [comment]
> Cons:
> 
>     * Not compatible with async (without relaxing the restriction to allow `ref struct`s to exist between await/yield points)

Hold on, isn't the `ref struct` prohibition in `async` context a language constraint, whereas the collection literal (expression) generation code is IL output, and therefore not bound by language constraints?

**Reply by CyrusNajmabadi** [reply]
> Hold on, isn't the ref struct prohibition in async context 

It's not a question about a ref struct in an async context.  It's about a ref-struct existing *across* an `await`.

**Reply by jnm2** [reply]
Meaning, there's an await inside the collection expression.

But isn't there currently also a technical limitation about lowering to ref struct codegen in async methods, or was that bridge already crossed with interpolated string handlers?

**Reply by TahirAhmadov** [reply]
> It's not a question about a ref struct in an async context. It's about a ref-struct existing _across_ an `await`.

He wrote "between await/yield points" which led me to believe he meant something like this:
```cs
async Task Foo()
{
  await Bar();
  // ...
  U8String text = [(byte)'#', (byte)':', ..payload];
  // ...
  await Baz();
}
```

**Reply by CyrusNajmabadi** [reply]
Between awaits should be fine. Across an await is a significant problem. 

We can always still make this work though by putting into heap storage, then copying into the span.  But we'd also likely want to warn in that case.

**Reply by 333fred** [reply]
>  was that bridge already crossed with interpolated string handlers?

That bridge was crossed by interpolated string handlers* (only usage of the DefaultInterpolatedStringHandler when lowering to a `string`), only because we have precise control over the lifetime and didn't need to introduce any new diagnostics for the scenario.
